rabit/rabit/Common/RealmManager.swift

```diff
@@ -3,30 +3,59 @@ import RealmSwift
 
 final class RealmManager {
     
-    private let realm: Realm
+    private let realmSerialQueue = DispatchQueue(label: "realm-serial-queue")
+    private let configuration = Realm.Configuration.defaultConfiguration
+    private let semaphore = DispatchSemaphore(value: 0)
     static var shared = RealmManager()
     
-    private init() {
-        self.realm = try! Realm()
-    }
-    
+    private init() { }
+        
     func read<T: Object>(entity: T.Type, filter query: String? = nil) -> [T] {
-        if let query = query {
-            return realm.objects(entity).filter(query).toArray(ofType: entity)
-        } else {
-            return realm.objects(entity).toArray(ofType: entity)
+        var result: [T] = []
+        
+        realmSerialQueue.async { [weak self] in
+            guard let self = self else {
+                self?.semaphore.signal()
+                return
+            }
+            
+            let realm = try! Realm(configuration: self.configuration, queue: self.realmSerialQueue).freeze()
+            
+            if let query = query {
+                result = realm.objects(entity).filter(query).toArray(ofType: entity)
+            } else {
+                result = realm.objects(entity).toArray(ofType: entity)
+            }
+            
+            self.semaphore.signal()
         }
+        
+        semaphore.wait()
```

Comment:
원래는 semaphore을 사용하지 않고 진행하려 했는데 ㅠㅠ serial queue로 Realm 생성 및 데이터 읽기 작업을 할당하다보니, 해당 작업이 끝날 때 까지 메소드가 리턴되지 않도록 대기해야 되서 우선은 부득이하게 사용했습니다.

다만... 해당 코드는 아직도 좋다고 생각되지 않는게 semaphore 혹은 lock의 잘못된 사용은 자칫 교착상태 혹은 병목을 유발할 수 있기 때문에 이런거 없이 가는게 궁극적으로 맞는 것 같아요..

또한 공식문서에서도 읽기작업을 할 때는 lock을 사용하지 말라고 권고하고 있네요 ....

> 공식문서 설명

### Don't lock to read:
Realm Database's [Multiversion Concurrency Control (MVCC)](https://www.mongodb.com/docs/realm/sdk/swift/crud/threading/#std-label-ios-mvcc) architecture eliminates the need to lock for read operations. The values you read will never be corrupted or in a partially-modified state. You can freely read from realms on any thread without the need for locks or mutexes. Unnecessary locking would be a performance bottleneck since each thread might need to wait its turn before reading.


`MVCC` 는 보통 데이터베이스 쪽에서 많이 사용되는 아키텍쳐 용어 같던데... 여기까지 이해하자니 저도 야크털 깎는게 아닌가 싶네요 ㅋㅋㅋㅋ

Reply:
쉽지 않네요.. 저는 처음에 Thread-safe하게 만드신다고 하셔서, 읽기에 대한 작업은 동시 액세스 허용/ 쓰기에 대한 작업은 배타적 액세스 형태의 Thread-Safe인줄 알았는데, Realm의 특성상 모두 똑같은 스레드에서 접근해야하는 문제가 있어 읽기/쓰기 모두 배타적 액세스를 해줘야하는거 같아서 더 복잡해지는거 같아요..

Reply:
넵넵ㅋㅋ Realm 자체가 뭔가 바뀔 때마다 여러 버전의 스냅샷을 트리구조로 저장해놓고, 최신 버전을 사용한다.. 뭐 이런거 같은데
그래서 하나의 Realm을 생성하면, 이후에는 무조건 생성 작업을 처리한 스레드에서만 접근해야 되고 그래서 굳이 일관성 유지를 위해 lock을 할 필요가 없다고 이해했어요! (써놓고도 제대로 이해가 안가네요 ㅎㅎ)

하나 더 흥미로운건... `여러 스냅샷을 저장` 하게 되면 파일 크기가 커지는거 아닌가? 하는 생각이 들었는데
(Realm 사용과 파일 크기)[https://www.mongodb.com/docs/realm/sdk/swift/realm-files/compacting/#std-label-ios-file-size] 와 같은 것도 공식문서에 있긴 하네요 ㅋㅋ